text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+The information necessary for restarting the network can (and probably should) be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Here we will just summarize the exact steps to be performed when an equivocation catastrophe happens.
+
+1. An equivocation catastrophe is detected. Once a single validator detects it, we can expect all the validators to detect it eventually as well (even if just due to them receiving the evidence from the validator that detected it first).
+2. Consensus gets suspended, though validators continue to gossip. The users (validator nodes' owners) receive notice about the catastrophe and start the process of social consensus.
+3. The users agree about the last correct block, the transactions that need to be replayed on top of it and the set of equivocating validators (exchanging evidence if necessary).
+4. A chainspec for the restart is manually constructed and agreed upon. This includes:
+    - the initial set of validators for the restarted network which excludes the equivocators,
+    - the initial global state hash for the restarted network
+5. The network is being restarted using the chainspec constructed above.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+The single drawback is that this proposal requires manual intervention in the case of an equivocation catastrophe. An ideal situation would be if the network could handle these automatically, but this is most probably impossible.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As mentioned above, automatic handling of equivocation catastrophes is probably impossible (as it would effectively mean constructing a consensus algorithm that exceeds proven theoretical limitations), so manual intervention is really the single option of dealing with such situations.
+
+An alternative to social consensus would be to have a designated party deal with the situation. However, such a solution would go against the goal of creating a truly distributed system. In this alternative, the party responsible for handling equivocation catastrophes would effectively single-handedly control the network, which is something we want to avoid.
+
+Not handling equivocation catastrophes at all would mean the end of the network once such a situation happens. Thus, even though we expect the probability of an equivocation catastrophe happening to be low (and we are making efforts towards making it as low as possible), the costs in the event of it happening are great enough to justify guarding against it.
+
+## Prior art
+
+[prior-art]: #prior-art
+
+There have been cases of blockchain projects introducing manual interventions in order to revert catastrophical events, for example the Ethereum DAO hack.
```

Comment:
It would be interesting to collect a few links to similar documentation of other projects. (But that would probably be a whole separate research story.)

You could link to [Vitalik's blog post](https://ethereum.github.io/blog/2014/11/25/proof-stake-learned-love-weak-subjectivity/), though.

Also, [this Confluence page](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/146047313/Weak+subjectivity) is probably obsolete now. Not sure whether we archive or delete those…?

Reply:
> similar documentation of other projects

Ah, that's probably covered by [RES-324](https://casperlabs.atlassian.net/browse/RES-324).

Reply:
@MParlikar, would Mallika have time to write something up for us before this CEP resolves? Re: RES-234

Reply:
Here is the research: https://drive.google.com/file/d/1XuZ_3IcYbhg9hkPw2h6HH5rdkBsf2JS-/view?usp=sharing